### DISCUSSION TITLE
Using git (JGit) for keeping history of persisted data

### DISCUSSION THREAD

**Discussion by chimp1984** [bisq-network/bisq2#8] [Ideas]
The rolling backups in Bisq are not working as well as hoped as it is hard to determine the right strategy when files get pruned. Using git instead might be an interesting approach.
There is a java library for git (https://www.eclipse.org/jgit/) which is just about 2.4 MB and covers the main git commands.
With that we could auto commit at each write operation and get a full backup of all changes. Not sure how large it would become but as it stores only the diffs it should be manageable.

**Comment by chimp1984** [comment]
Seems to be not a good idea. Tried it with Bisq running one night and was from initial 250 MB growing to 1 GB...
We need a smarter rolling backup strategy similar to time machine in OSX where one get fine grained snapshots more recently and less frequent snapshots when further back in time.
Anyone aware of such a library in java?

**Reply by wiz** [reply]
How about simply keep the last 10 files or so?

**Reply by chimp1984** [reply]
Thats what we do now. Depending on importance of the file x files are kept, similar like rolling logs.
But the files are updated each write operation so it can be that the oldest file is still pretty young if there have been more write activity. When there have been cases where backup files have been important I guess 50% of the cases the files have not been there anymore. So the strategy is not super useful in reality.
Timemachine does it pretty good. You get fine grained history for short term past and it gets more crouse when you get back in time, e.g. after 3 months you only have 1 backup per day... I think something similar would be helpful.

**Reply by cd2357** [reply]
Written in C, but https://github.com/borgbackup/borg can do that.

From their [documentation](https://borgbackup.readthedocs.io/en/stable/usage/prune.html#):
```
# Keep 7 end of day and 4 additional end of week archives.
# Do a dry-run without actually deleting anything.
$ borg prune -v --list --dry-run --keep-daily=7 --keep-weekly=4 /path/to/repo

# Same as above but only apply to archive names starting with the hostname
# of the machine followed by a "-" character:
$ borg prune -v --list --keep-daily=7 --keep-weekly=4 --prefix='{hostname}-' /path/to/repo

# Keep 7 end of day, 4 additional end of week archives,
# and an end of month archive for every month:
$ borg prune -v --list --keep-daily=7 --keep-weekly=4 --keep-monthly=-1 /path/to/repo

# Keep all backups in the last 10 days, 4 additional end of week archives,
# and an end of month archive for every month:
$ borg prune -v --list --keep-within=10d --keep-weekly=4 --keep-monthly=-1 /path/to/repo
```

If the binary is on the PATH, it could be called from within Misq. It could be a pre-requisite that users have it on their system.

**Reply by cd2357** [reply]
However the problem with persisting files is that we want atomic snapshots. A write operation could affect multiple files, but if one of them fails to write, restoring that snapshot results in an inconsistent state. Unless we persist everything in one file, the only way to make it atomic is to use some sort of DB engine (like H2) with built-in atomic transactions.

**Reply by chimp1984** [reply]
I think that would be too much of a dependency. At the end backup is users responsibility anyway, other software usually dont offer it as well. In Bisq it got added as rolling backups was so easy...

We do write to temp file then rename, so its atomic as rename is atomic. If write fails the temp file is affected only which gets deleted on exit.
We also store each data type in a file (e.g. similar as in Bisq but more fine grained).

The reason why I don't want to use a DB is because we do not do any queries and we read only at startup. Also I want to avoid that we need a second schema (beside protobuf). So the only advantage of a DB would be that it does the write operations, but at the cost of more overhead from storage file size, startup and initialisation and dependencies.
